logconsumer_test.go

```diff
@@ -508,3 +509,117 @@ func Test_StartLogProductionStillStartsWithTooHighTimeout(t *testing.T) {
 
 	terminateContainerOnEnd(t, ctx, c)
 }
+
+func Test_MultiContainerLogConsumer_CancelledContext(t *testing.T) {
+	// Redirect stderr to a buffer
+	oldStderr := os.Stderr
+	r, w, _ := os.Pipe()
+	os.Stderr = w
+
+	// Context with cancellation functionality for simulating user interruption
+	ctx, cancel := context.WithCancel(context.Background())
+
+	first := TestLogConsumer{
+		Msgs:     []string{},
+		Done:     make(chan bool),
+		Accepted: devNullAcceptorChan(),
+	}
+
+	containerReq1 := ContainerRequest{
+		FromDockerfile: FromDockerfile{
+			Context:    "./testdata/",
+			Dockerfile: "echoserver.Dockerfile",
+		},
+		ExposedPorts: []string{"8080/tcp"},
+		WaitingFor:   wait.ForLog("ready"),
+		LogConsumerCfg: &LogConsumerConfig{
+			Consumers: []LogConsumer{&first},
+		},
+	}
+
+	genericReq1 := GenericContainerRequest{
+		ContainerRequest: containerReq1,
+		Started:          true,
+	}
+
+	c, err := GenericContainer(ctx, genericReq1)
+	require.NoError(t, err)
+
+	ep1, err := c.Endpoint(ctx, "http")
+	require.NoError(t, err)
+
+	_, err = http.Get(ep1 + "/stdout?echo=hello1")
+	require.NoError(t, err)
+
+	_, err = http.Get(ep1 + "/stdout?echo=there1")
+	require.NoError(t, err)
+
+	second := TestLogConsumer{
+		Msgs:     []string{},
+		Done:     make(chan bool),
+		Accepted: devNullAcceptorChan(),
+	}
+
+	containerReq2 := ContainerRequest{
+		FromDockerfile: FromDockerfile{
+			Context:    "./testdata/",
+			Dockerfile: "echoserver.Dockerfile",
+		},
+		ExposedPorts: []string{"8080/tcp"},
+		WaitingFor:   wait.ForLog("ready"),
+		LogConsumerCfg: &LogConsumerConfig{
+			Consumers: []LogConsumer{&second},
+		},
+	}
+
+	genericReq2 := GenericContainerRequest{
+		ContainerRequest: containerReq2,
+		Started:          true,
+	}
+
+	c2, err := GenericContainer(ctx, genericReq2)
+	require.NoError(t, err)
+
+	ep2, err := c2.Endpoint(ctx, "http")
+	require.NoError(t, err)
+
+	_, err = http.Get(ep2 + "/stdout?echo=hello2")
+	require.NoError(t, err)
+
+	_, err = http.Get(ep2 + "/stdout?echo=there2")
+	require.NoError(t, err)
+
+	// Handling the termination of the containers
+	defer func() {
+		shutdownCtx, shutdownCancel := context.WithTimeout(
+			context.Background(), 60*time.Second,
+		)
+		defer shutdownCancel()
+		_ = c.Terminate(shutdownCtx)
+		_ = c2.Terminate(shutdownCtx)
+	}()
+
+	// Deliberately calling context cancel
+	cancel()
+
+	// We check log size due to context cancellation causing
+	// varying message counts, leading to test failure.
+	assert.GreaterOrEqual(t, len(first.Msgs), 2)
+	assert.GreaterOrEqual(t, len(second.Msgs), 2)
```

Comment:
I think the PR is good to me, although I'd like to understand more on this: we are adding here a wider way to capture the messages. Could you elaborate on this a little bit?

Reply:
@mdelapenya : Sure, the main reason for checking the message length was due to the flaky nature of the test. Using `cancel()` just above the check is causing the result to show only "ready" and "echo hello1" messages. In some cases, it's not showing the last message "echo there1," which leads to test failures. Please let me know if you'd like further elaboration.

Reply:
@mdelapenya : All the checks seem to have passed. If you wouldn't mind taking a final look and are comfortable with it, merging it would be a big help. Our development process relies on this fix. Thanks!

Reply:
Hi @prateekdwivedi will do the review today, thanks for your patience 🙇 

> Our development process relies on this fix. Thanks!

You can always add a replace condition in your projects targeting your branch, and then replace it once we release it

Reply:
Thank you so much @mdelapenya

Reply:
@mdelapenya : Please let me know if you have any further concerns. I totally understand that you must be swamped with a ton of things, so please feel free to respond whenever it suits you best.

Reply:
Thank you for reminding it 🙇 I do appreciate your time and I will be on this PR now. Sorry for the delay, many times jumping from one place to the other... 🤦